File: LeetCode/p02_add_two_numbers/addTwoNumbers.ts
// class ListNode {
//     val: number
//     next: ListNode | null
//     constructor(val?: number, next?: ListNode | null) {
//         this.val = (val===undefined ? 0 : val)
//         this.next = (next===undefined ? null : next)
//     }
// }

function addTwoNumber(l1: ListNode | null, l2: ListNode | null,): ListNode | null{
    let head = new ListNode(0);
    let current = head;
    let carry = 0;

    while(l1 != null || l2 != null || carry != 0){
        let l1Val = l1 != null ? l1.val : 0;
        let l2Val = l2 != null ? l2.val : 0;
        
        let sum = carry + l1Val + l2Val;
        carry = Math.floor(sum / 10);
        current.next = new ListNode(sum % 10);
        current = current.next;
        
        if(l1 !== null) l1 = l1.next;
        if(l2 !== null) l2 = l2.next;
    }
    return head.next;
}

File: LeetCode/p02_add_two_numbers/addTwoNumbers.kt
//  var li = ListNode(5)
//  var v = li.`val`
//
//  class ListNode(var `val`: Int) {
//      var next: ListNode? = null
//  }

class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        val head = ListNode(0)
        var current = head

        var p1 = l1
        var p2 = l2
        var carry = 0

        while(p1 != null || p2 != null || carry != 0) {
            val x = p1?.`val` ?: 0
            val y = p2?.`val` ?: 0

            val sum = carry + x + y
            carry = sum / 10

            current.next = ListNode(sum % 10)
            current = current.next!!

            p1 = p1?.next
            p2 = p2?.next
        }
        return head.next
    }
}